grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/GreedyTraversalPlan.java

```diff
@@ -136,6 +174,123 @@ public static GraqlTraversal createTraversal(PatternAdmin pattern, EmbeddedGrakn
+        allFragments.stream().filter(OutRolePlayerFragment.class::isInstance)
+                .forEach(fragment -> relationshipRolePlayerMap.put(fragment.start(), fragment.end()));
+
+        // find all the relationships requiring type inference
+        Iterator<Var> iterator = relationshipRolePlayerMap.keySet().iterator();
+        while (iterator.hasNext()) {
+            Var relationship = iterator.next();
+
+            // the relation should have at least 2 known role players so we can infer something useful
+            if (instanceVarTypeMap.containsKey(relationship) ||
+                    relationshipRolePlayerMap.get(relationship).size() < 2) {
+                iterator.remove();
+            } else {
+                final int[] numRolePlayersHaveType = {0};
+                relationshipRolePlayerMap.get(relationship).forEach(
+                        rolePlayer -> {
+                            if (instanceVarTypeMap.containsKey(rolePlayer)) {
+                                numRolePlayersHaveType[0]++;
+                            }
+                        });
+                if (numRolePlayersHaveType[0] < 2) {
+                    iterator.remove();
+                }
+            }
+        }
+        if (relationshipRolePlayerMap.isEmpty()) return;
+
+        // for each type, get all possible relationship type it could be in
+        Map<Type, Set<RelationshipType>> relationshipMap = new HashMap<>();
+        getAllPossibleRelationships(relationshipMap, tx.admin().getMetaEntityType());
+        getAllPossibleRelationships(relationshipMap, tx.admin().getMetaAttributeType());
+        getAllPossibleRelationships(relationshipMap, tx.admin().getMetaRelationType());
+
+        relationshipRolePlayerMap.asMap().forEach((relationshipVar, rolePlayerVars) -> {
+            Iterator<Var> rolePlayerVarIterator = rolePlayerVars.iterator();
+
+            Var firstRolePlayer = rolePlayerVarIterator.next();
+            while (!instanceVarTypeMap.containsKey(firstRolePlayer)) {
+                firstRolePlayer = rolePlayerVarIterator.next();
+            }
+
+            Set<Type> possibleRelationShipTypes = getAllPossibleRelationShipTypesOfARolePlayer(
+                    firstRolePlayer, instanceVarTypeMap, relationshipMap);
+            while (rolePlayerVarIterator.hasNext()) {
+                Var nextRolePlayer = rolePlayerVarIterator.next();
+                if (instanceVarTypeMap.containsKey(nextRolePlayer)) {
+
+                    // compute the intersection of possible relationship types
+                    possibleRelationShipTypes.retainAll(getAllPossibleRelationShipTypesOfARolePlayer(
+                            nextRolePlayer, instanceVarTypeMap, relationshipMap));
+                }
+            }
+
+            //TODO: if possibleRelationShipTypes here is empty, the query will not match any data
+            if (possibleRelationShipTypes.size() == 1) {
```

Comment:
What if we end up with have more than 1 type but all types are in sub hierarchy?

Reply:
We do nothing, for now :-)